### ISSUE TITLE
wantarray() causes clobbering of unrelated vars outside the sub

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5683]
Migrated from [rt.perl.org#9998](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=9998) (status was 'resolved')

Searchable as RT9998$

**Comment by p5pRT** [comment]
### From ethan@r220-1.rz.RWTH-Aachen.DE

#### Created by tassilo.parseval@post.rwth-aachen.de



The following code produces unexplicable behaviour\. It can probably be pinned
down to wantarray\(\)\, but obviously other factors are playing a role as well
\(notably keys\(\)\)&#8203;:


\#\! /usr/bin/perl

use strict;
use warnings;

my %tmp\_hash = \( first => 'A'\, second => 'B' \);

foreach my $aaa \( keys %tmp\_hash \) \{
&nbsp;    print "$aaa \-\-> $tmp\_hash\{ $aaa \}\\n";
\}

print "First loop&#8203;:\\n";
for my $bbb \( keys %tmp\_hash \) \{
&nbsp;    print "key1&#8203;: $bbb\\n";
&nbsp;    my \($jjj\) = mysub\(\);
&nbsp;    print "key2&#8203;: $bbb\\n";
\}

print "Second loop&#8203;:\\n";
my @&#8203;arr = keys %tmp\_hash;
foreach my $bbb \( @&#8203;arr \) \{
&nbsp;    print "key3&#8203;: $bbb\\n";
&nbsp;    my \( $jjj \) = mysub\(\);
&nbsp;    print "key4&#8203;: $bbb\\n";
\}

sub mysub \{
&nbsp;    my $text = 'select \* from employee';
&nbsp;    my $val;
&nbsp;    if \( wantarray and $text =~ /^update/ \) \{
&nbsp;        $text;
&nbsp;    \}
\}

Output&#8203;:
\-\-\-\-\-\-\-
first \-\-> A
second \-\-> B
First loop&#8203;:
key1&#8203;: first
key2&#8203;: first
Use of uninitialized value in concatenation \(\.\) or string at test\.pl line 14\.
key1&#8203;:
key2&#8203;: select \* from employee
Second loop&#8203;:
key3&#8203;: first
key4&#8203;: first
key3&#8203;: second
key4&#8203;: second
\-\-\-\-\-\-\-\-

The warning of unitialized value is spurious here\. Furthermore\, the second
iteration through the First\-Loop\-foreach goes terribly wrong\. Calling mysub\(\)
in list context changes the content of the lexical $bbb\. Interestingly enough\,
this does not happen in the second loop that iterates over a copy of keys\(\) and
not over the keys directly\.

According to Slaven Rezic\, this bug occurs beginning with 5\.00503 whereas it
works fine with perl5\.00404\. I personally observed the faulty behaviour in
5\.8\.0rc2 and 5\.6\.1\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.0:

Configured by root at Mon Jun 24 08:29:49 CEST 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.18, archname=i686-linux
    uname='linux ethan 2.4.18 #1 tue feb 26 09:52:36 cet 2002 i686 unknown '
    config_args='-e -Dprefix=/usr/opt'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.4 20011006 (Debian prerelease)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldbm -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.5.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/opt/lib/perl5/5.8.0/i686-linux
    /usr/opt/lib/perl5/5.8.0
    /usr/opt/lib/perl5/site_perl/5.8.0/i686-linux
    /usr/opt/lib/perl5/site_perl/5.8.0
    /usr/opt/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/ethan
    LANG=C
    LANGUAGE (unset)
    LC_CTYPE=de_DE@euro
    LD_LIBRARY_PATH=:/usr/local/j2sdk1.4.0/jre/lib/i386:/usr/local/pwlib/lib
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/bin:/bin:/usr/X11R6/bin:/usr/local/jdk1.3.1/bin:.
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Jun 26\, 2002 at 08&#8203;:03&#8203;:13PM \+0200\, Tassilo von Parseval wrote&#8203;:
> The following code produces unexplicable behaviour\. It can probably be pinned
> down to wantarray\(\)\, but obviously other factors are playing a role as well
> \(notably keys\(\)\)&#8203;:
> 

\[snip\]
>     if \( wantarray and $text =~ /^update/ \) \{

It appears to be an interaction between the 'wantarray'\,  'and' and 'match'
ops\.

When this boolean expression is executed\, for some reason the 'match' op
doesn't push an SV\_NO onto the stack\, so the following 'and' op eats into
the previous stack frame and everything goes bang\. For some reason this
only happens when the LHS of the 'and' is wantarray \- otherwise 'match'
behaves okay\. I need some sleep now\, so if anyone else wants to look at it
in the meantime\, feel free\.


&nbsp;    =>  PVIV\("first"\)  PVIV\("second"\)  \*  
\(/tmp/p1&#8203;:27\)    wantarray
&nbsp;    =>  PVIV\("first"\)  PVIV\("second"\)  \*  SV\_YES  
\(/tmp/p1&#8203;:27\)    and
&nbsp;    =>  PVIV\("first"\)  PVIV\("second"\)  \*  
\(/tmp/p1&#8203;:27\)    padsv\($text\)
&nbsp;    =>  PVIV\("first"\)  PVIV\("second"\)  \*  PV\("select \* from employee"\\0\)  
\(/tmp/p1&#8203;:27\)    match
&nbsp;    =>  PVIV\("first"\)  PVIV\("second"\)  \*  \<==== XXX no result pushed
\(/tmp/p1&#8203;:27\)    and
&nbsp;    =>  PVIV\("first"\)                     \<==== bang goes the stack


\-\- 
"But Sidley Park is already a picture\, and a most amiable picture too\.
The slopes are green and gentle\. The trees are companionably grouped at
intervals that show them to advantage\. The rill is a serpentine ribbon
unwound from the lake peaceably contained by meadows on which the right
amount of sheep are tastefully arranged\." Lady Croom \- Arcadia

**Comment by p5pRT** [comment]
### From @vanstyn

Dave Mitchell \<davem@&#8203;fdgroup\.com> wrote&#8203;:
:On Wed\, Jun 26\, 2002 at 08&#8203;:03&#8203;:13PM \+0200\, Tassilo von Parseval wrote&#8203;:
:> The following code produces unexplicable behaviour\. It can probably be pinned
:> down to wantarray\(\)\, but obviously other factors are playing a role as well
:> \(notably keys\(\)\)&#8203;:
:> 
:
:\[snip\]
:>     if \( wantarray and $text =~ /^update/ \) \{
:
:It appears to be an interaction between the 'wantarray'\,  'and' and 'match'
:ops\.

It appears to be an optree problem\. An initial look suggests that the
two cases where we handle OP\_WANTARRAY in op\.c \- inside newCONDOP\(\) and
new\_logop\(\) \- we try to propagate the appropriate context to the child
ops by assuming that the 'wantarray' branch should have list context
and the 'not wantarray' branch should have scalar context\. However this
does not take into account the surrounding context; in this case&#8203;:
&nbsp;  sub a \{ wantarray ? 1 : 2 \}
the '1' branch should be in list context\, but in this case&#8203;:
&nbsp;  sub a \{ $x = wantarray ? 1 : 2 \}
the '1' branch should be in scalar context\.

Patch below marks the points\. I'm not sure\, but I think a correct
solution to this involves removing these blocks as premature\, and
instead having ck\_return\(\) propagate a 'varying' context that
would catch the opportunity to handle OP\_WANTARRAY in these
situations; cases where wantarray is not used in a returned context
would be caught by the normal list\(\) or scalar\(\) propagation\.

Hugo
<details><summary>Inline Patch</summary>

```diff
--- op.c.old	Sat Jun 15 17:43:11 2002
+++ op.c	Thu Jun 27 11:42:15 2002
@@ -3900,6 +3900,8 @@
 	}
     }
     else if (first->op_type == OP_WANTARRAY) {
+	/* XXX true only if this result will be returned, else should
+	   propagate outer context */
 	if (type == OP_AND)
 	    list(other);
 	else
@@ -3995,6 +3997,8 @@
 	}
     }
     else if (first->op_type == OP_WANTARRAY) {
+	/* XXX true only if this result will be returned, else should
+	   propagate outer context */
 	list(trueop);
 	scalar(falseop);
     }
```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

On 2002\.06\.27 12&#8203;:58 Hugo van der Sanden wrote&#8203;:
> 
> Patch below marks the points\.

Points marked \(\#17369\)\.

**Comment by p5pRT** [comment]
### From @ysth

In article \<20020627014411\.A5285@&#8203;fdgroup\.com>\,
Dave Mitchell \<davem@&#8203;fdgroup\.com> wrote&#8203;:
>On Wed\, Jun 26\, 2002 at 08&#8203;:03&#8203;:13PM \+0200\, Tassilo von Parseval wrote&#8203;:
>> The following code produces unexplicable behaviour\. It can probably be pinned
>> down to wantarray\(\)\, but obviously other factors are playing a role as well
>> \(notably keys\(\)\)&#8203;:
>> 
>
>\[snip\]
>>     if \( wantarray and $text =~ /^update/ \) \{
>
>It appears to be an interaction between the 'wantarray'\,  'and' and 'match'
>ops\.
>
>When this boolean expression is executed\, for some reason the 'match' op
>doesn't push an SV\_NO onto the stack\, so the following 'and' op eats into
>the previous stack frame and everything goes bang\. For some reason this
>only happens when the LHS of the 'and' is wantarray \- otherwise 'match'
>behaves okay\. I need some sleep now\, so if anyone else wants to look at it
>in the meantime\, feel free\.

Look for OP\_WANTARRAY in op\.c\.  Both places we impose context based on
it are broken\.  They blindly assume the result of the logop/condop is
only being returned by the sub\.

**Comment by p5pRT** [comment]
### From @vanstyn

Earlier I wrote&#8203;:
:Dave Mitchell \<davem@&#8203;fdgroup\.com> wrote&#8203;:
:&#8203;:On Wed\, Jun 26\, 2002 at 08&#8203;:03&#8203;:13PM \+0200\, Tassilo von Parseval wrote&#8203;:
:&#8203;:> The following code produces unexplicable behaviour\. It can probably be pinned
:&#8203;:> down to wantarray\(\)\, but obviously other factors are playing a role as well
:&#8203;:> \(notably keys\(\)\)&#8203;:
:&#8203;:> 
:&#8203;:
:&#8203;:\[snip\]
:&#8203;:>     if \( wantarray and $text =~ /^update/ \) \{
:&#8203;:
:&#8203;:It appears to be an interaction between the 'wantarray'\,  'and' and 'match'
:&#8203;:ops\.
:
:It appears to be an optree problem\. An initial look suggests that the
:two cases where we handle OP\_WANTARRAY in op\.c \- inside newCONDOP\(\) and
:new\_logop\(\) \- we try to propagate the appropriate context to the child
:ops by assuming that the 'wantarray' branch should have list context
:and the 'not wantarray' branch should have scalar context\. However this
:does not take into account the surrounding context; in this case&#8203;:
:  sub a \{ wantarray ? 1 : 2 \}
:the '1' branch should be in list context\, but in this case&#8203;:
:  sub a \{ $x = wantarray ? 1 : 2 \}
:the '1' branch should be in scalar context\.

I originally attempted to fix this simple by removing the relevant
branches as undesired \(and incorrect\) optimisations\, but had nasty
problems early in the build cycle that I assumed meant it was nowhere
near that simple\.

Looking at that nasty problem again\, however\, I've come to the
conclusion that it was simply a bug in the perl code that had been
masked by this optimisation\. So I think the patch below is a correct
fix\.

Hugo
<details><summary>Inline Patch</summary>

```diff
--- op.c.old	Sun Jul  7 19:10:06 2002
+++ op.c	Mon Jul  8 16:46:45 2002
@@ -3899,14 +3899,6 @@
 	    return first;
 	}
     }
-    else if (first->op_type == OP_WANTARRAY) {
-	/* XXX true only if this result will be returned, else should
-	   propagate outer context */
-	if (type == OP_AND)
-	    list(other);
-	else
-	    scalar(other);
-    }
     else if (ckWARN(WARN_MISC) && (first->op_flags & OPf_KIDS)) {
 	OP *k1 = ((UNOP*)first)->op_first;
 	OP *k2 = k1->op_sibling;
@@ -3995,12 +3987,6 @@
 	    op_free(trueop);
 	    return falseop;
 	}
-    }
-    else if (first->op_type == OP_WANTARRAY) {
-	/* XXX true only if this result will be returned, else should
-	   propagate outer context */
-	list(trueop);
-	scalar(falseop);
     }
     NewOp(1101, logop, 1, LOGOP);
     logop->op_type = OP_COND_EXPR;
--- lib/File/Basename.pm.old	Mon Jul  8 15:16:10 2002
+++ lib/File/Basename.pm	Mon Jul  8 15:16:37 2002
@@ -226,7 +226,7 @@
 
   $tail .= $taint if defined $tail; # avoid warning if $tail == undef
   wantarray ? ($basename .= $taint, $dirpath .= $taint, $tail)
-            : $basename .= $taint;
+            : ($basename .= $taint);
 }
 
 
--- t/op/wantarray.t.old	Tue Aug  1 03:32:14 2000
+++ t/op/wantarray.t	Mon Jul  8 16:47:02 2002
@@ -1,6 +1,6 @@
 #!./perl
 
-print "1..7\n";
+print "1..9\n";
 sub context {
   my ( $cona, $testnum ) = @_;
   my $conb = (defined wantarray) ? ( wantarray ? 'A' : 'S' ) : 'V';
@@ -17,4 +17,18 @@
 $a = scalar context('S',5);
 ($a) = context('A',6);
 ($a) = scalar context('S',7);
+
+{
+  # [ID 20020626.011] incorrect wantarray optimisation
+  sub simple { wantarray ? 1 : 2 }
+  sub inline {
+    my $a = wantarray ? simple() : simple();
+    $a;
+  }
+  my @b = inline();
+  my $c = inline();
+  print +(@b == 1 && "@b" eq "2") ? "ok 8\n" : "not ok 8\t# <@b>\n";
+  print +($c == 2) ? "ok 9\n" : "not ok 9\t# <$c>\n";
+}
+
 1;
```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

On 2002\.07\.08 18&#8203;:00 Hugo van der Sanden wrote&#8203;:
> 
> Looking at that nasty problem again\, however\, I've come to the
> conclusion that it was simply a bug in the perl code that had been
> masked by this optimisation\. So I think the patch below is a correct
> fix\.

In any case the File&#8203;::Basename patch is needed\.

**Comment by p5pRT** [comment]
### From @jhi

> I originally attempted to fix this simple by removing the relevant
> branches as undesired \(and incorrect\) optimisations\, but had nasty
> problems early in the build cycle that I assumed meant it was nowhere
> near that simple\.
> 
> Looking at that nasty problem again\, however\, I've come to the
> conclusion that it was simply a bug in the perl code that had been
> masked by this optimisation\. So I think the patch below is a correct
> fix\.

Thanks\, applied \(\#17423\)\.

\-\- 
$jhi\+\+; \# http&#8203;://www\.iki\.fi/jhi/
&nbsp;        \# There is this special biologist word we use for 'stable'\.
&nbsp;        \# It is 'dead'\. \-\- Jack Cohen

**Comment by p5pRT** [comment]
### From @ysth

In article \<20020709041252\.E16125@&#8203;alpha\.hut\.fi>\,
Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> wrote&#8203;:
>> I originally attempted to fix this simple by removing the relevant
>> branches as undesired \(and incorrect\) optimisations\, but had nasty
>> problems early in the build cycle that I assumed meant it was nowhere
>> near that simple\.
>> 
>> Looking at that nasty problem again\, however\, I've come to the
>> conclusion that it was simply a bug in the perl code that had been
>> masked by this optimisation\. So I think the patch below is a correct
>> fix\.
>
>Thanks\, applied \(\#17423\)\.

Bug admins note that this should also fix 20020623\.020\.

**Comment by p5pRT** [comment]
### From @nwc10

On Mon\, Jul 08\, 2002 at 09&#8203;:37&#8203;:32PM \-0700\, Yitzchak Scott\-Thoennes wrote&#8203;:
> In article \<20020709041252\.E16125@&#8203;alpha\.hut\.fi>\,
> Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> wrote&#8203;:
> >> I originally attempted to fix this simple by removing the relevant
> >> branches as undesired \(and incorrect\) optimisations\, but had nasty
> >> problems early in the build cycle that I assumed meant it was nowhere
> >> near that simple\.
> >> 
> >> Looking at that nasty problem again\, however\, I've come to the
> >> conclusion that it was simply a bug in the perl code that had been
> >> masked by this optimisation\. So I think the patch below is a correct
> >> fix\.
> >
> >Thanks\, applied \(\#17423\)\.
> 
> Bug admins note that this should also fix 20020623\.020\.

It does seem to&#8203;:

17403&#8203;:
$ \./perl
sub foo \{my @&#8203;x=wantarray?bar\(\)&#8203;:bar\(\)\}; \# Impose list context on bar
sub bar \{wantarray?print"list\\n"&#8203;:print"scalar\\n"\}; \# Report context
my $a=foo\(\); \# bar prints "scalar"\.
scalar


17429&#8203;:
$ \./perl
sub foo \{my @&#8203;x=wantarray?bar\(\)&#8203;:bar\(\)\}; \# Impose list context on bar
sub bar \{wantarray?print"list\\n"&#8203;:print"scalar\\n"\}; \# Report context
my $a=foo\(\); \# bar prints "scalar"\.
list

I've marked it as closed\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @ysth

In article \<20020709112002\.G5635@&#8203;plum\.flirble\.org>\,
Nicholas Clark \<nick@&#8203;ccl4\.org> wrote&#8203;:
>On Mon\, Jul 08\, 2002 at 09&#8203;:37&#8203;:32PM \-0700\, Yitzchak Scott\-Thoennes wrote&#8203;:
>> In article \<20020709041252\.E16125@&#8203;alpha\.hut\.fi>\,
>> Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> wrote&#8203;:
>> >> I originally attempted to fix this simple by removing the relevant
>> >> branches as undesired \(and incorrect\) optimisations\, but had nasty
>> >> problems early in the build cycle that I assumed meant it was nowhere
>> >> near that simple\.
>> >> 
>> >> Looking at that nasty problem again\, however\, I've come to the
>> >> conclusion that it was simply a bug in the perl code that had been
>> >> masked by this optimisation\. So I think the patch below is a correct
>> >> fix\.
>> >
>> >Thanks\, applied \(\#17423\)\.
>> 
>> Bug admins note that this should also fix 20020623\.020\.
>
>It does seem to&#8203;:
>
>17403&#8203;:
>$ \./perl
>sub foo \{my @&#8203;x=wantarray?bar\(\)&#8203;:bar\(\)\}; \# Impose list context on bar
>sub bar \{wantarray?print"list\\n"&#8203;:print"scalar\\n"\}; \# Report context
>my $a=foo\(\); \# bar prints "scalar"\.
>scalar
>
>
>17429&#8203;:
>$ \./perl
>sub foo \{my @&#8203;x=wantarray?bar\(\)&#8203;:bar\(\)\}; \# Impose list context on bar
>sub bar \{wantarray?print"list\\n"&#8203;:print"scalar\\n"\}; \# Report context
>my $a=foo\(\); \# bar prints "scalar"\.
>list

I'd like to know when the offending sections of op\.c were added\, and why\.
My guess is that it was to trigger some desired compile\-time warning\.

Andreas\, can you determine when the above \*started\* printing 'scalar'?

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Wed\, 10 Jul 2002 18&#8203;:18&#8203;:34 \-0700\, sthoenna@&#8203;efn\.org \(Yitzchak Scott\-Thoennes\) said&#8203;:

 >> 17403&#8203;:
 >> $ \./perl
 >> sub foo \{my @&#8203;x=wantarray?bar\(\)&#8203;:bar\(\)\}; \# Impose list context on bar
 >> sub bar \{wantarray?print"list\\n"&#8203;:print"scalar\\n"\}; \# Report context
 >> my $a=foo\(\); \# bar prints "scalar"\.
 >> scalar
 >> 
 >> 
 >> 17429&#8203;:
 >> $ \./perl
 >> sub foo \{my @&#8203;x=wantarray?bar\(\)&#8203;:bar\(\)\}; \# Impose list context on bar
 >> sub bar \{wantarray?print"list\\n"&#8203;:print"scalar\\n"\}; \# Report context
 >> my $a=foo\(\); \# bar prints "scalar"\.
 >> list

&nbsp;  > I'd like to know when the offending sections of op\.c were added\, and why\.
&nbsp;  > My guess is that it was to trigger some desired compile\-time warning\.

&nbsp;  > Andreas\, can you determine when the above \*started\* printing 'scalar'?

Seems like it always has printed "scalar"\. My collection is complete
from 5\.005\_57\. I have only few older perls\, oldest being 5\.003\_07\. All
versions I have print "scalar"\.

\-\- 
andreas
